### DISCUSSION TITLE
Help: making a catppuccin color theme for ble.sh

### DISCUSSION THREAD

**Discussion by kjurl** [akinomyoga/ble.sh#411] [Q&A]
Setup: Using [WezTerm](https://wezfurlong.org/wezterm/index.html), with the [catppuccing theme](https://wezfurlong.org/wezterm/colorschemes/c/index.html?h=#catppuccin-mocha) enabled.

I've read the docs, seems like I need to create `~/.blerc` file containing a modified version of [default colors](https://github.com/akinomyoga/ble.sh?tab=readme-ov-file#26-highlight-colors), using the [wikia](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A72-Graphics) I found that colors could be defined in a number of ways: integer, a hexadecimal value, and a color name.

I want to know how the color name - `red`, `black`, `purple` .. etc work.

I initially thought just replacing default colors with their look-a-like counterparts in the [catppuccin palette](https://github.com/catppuccin/catppuccin?tab=readme-ov-file#-palette) would work. Upon further tinkering I found that ["faces"](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A72-Graphics#23-basic-faces) specified with color `red` have automatically adapted to the catppuccin version of `red`(#f38ba8).

![image](https://github.com/akinomyoga/ble.sh/assets/89933773/d183c1ee-c88e-4282-ae13-5ae8a5d23127)

Questions I had in my head-
1. How do these color names work? 
2. Where are they fetched from? 
3. Is is safe to assume that other catppuccin colors like `rosewater`(#f5e0dc) also do exist? (if so then developing a theme would be extremely simple.) 
4. Are these terminal dependent (like wezterm has catppuccin enabled, but say gnome-terminal doesnt)?
5. Can I add to this list of colors?

**Comment by akinomyoga** [comment]
### Color specifications in terminals

First I need to explain how the colors are specified in terminals. The colors are specified by the terminal applications (CUI commands and applications such as line editors like ble.sh and CUI text editors like Vim), and the terminals receive the color specifications and use them to render texts in the display. There are mainly two ways of the color specification. One is the index color, and the other is the direct specification.

For the index colors, the terminals have their own *palette*, where an index is assigned to each color. The palette depends on the terminal and its theme, but there is a convention in the correspondence between the colors and the indices. That is, index 0 is black(ish) color, index 1 is red(ish), index 2 is green(ish), index 3 is yellow(ish), index 4 is blue(ish), index 5 is magent(ish), index 6 is cyan(ish), and index 7 is whit(ish) color. Index 8..15 are brighter versions of index 0..7. In total, there are 16 index colors that usually depend on the terminal's *theme*. Note that it is just a *convention*, so a terminal can have completely different palette not following the convention. For example, the Solarized theme for terminals do not follow this convention. In the market, there is additional colors, index 16..255, for which there is a *de facto* standard by XTerm, and we can assume the palette is common for all the modern terminals.

For the index colors, the terminal applications specify just an index *without knowing how the index is exactly mapped to the actual color*. Nevertheless, the terminal applications can assume the *convention* for the first 16 colors (index 0..15).

Some of modern terminals also support the direct specification of the color using the 24-bit RGB values. 

### Color specification in ble.sh

The color names such as `red` (index 9), `black` (index 0), `purple` (index 5), etc. in ble.sh are actually index colors and named after *the convention* for the first 16 colors (except `orange` which is index 202). When those color names are used, `ble.sh` just sends "the index" to the terminal, and the terminal converts the index to the actual color based on its theme.

`ble.sh` also supports the direct color specification, which is used when the user specifies the color by the hex representation or other representations with color spaces. In this case, the terminal just uses the color directly specified by `ble.sh`.

### Answers to the question

> 1. How do these color names work?

As explained above.

> 2. Where are they fetched from?

`ble.sh` doesn't know the actual colors of the index colors. `ble.sh` just sends the corresponding index to the terminal. The terminal (wezterm in your case) converts the index to the actual color using its theme. In this sense, *wezterm fetches the color* from its theme and renders a text in the display using the obtained color.

> 3. Is is safe to assume that other catppuccin colors like rosewater(#f5e0dc) also do exist?

No. Those theme-dependent 16 colors are formed as a consensus among historical terminals and terminal applications, so one cannot arbitrarily add new theme-dependent colors. It's technically possible to define a new index for a specific color, but that breaks the assumptions of other terminal applications and might cause unexpected behavior (though it's just strange coloring at worst).

> 4. Are these terminal dependent (like wezterm has catppuccin enabled, but say gnome-terminal doesn't)?

The content of the palette is terminal-dependent and theme-dependent. However, the mechanism of the index colors are common in all the terminals. Rather, the direct color specifications are not supported by all the terminals.

> 5. Can I add to this list of colors?

As in an answer to Q3, there are no other theme-dependent colors, but we may still give names to the other "fixed" colors. That is, maybe we can introduce custom color mapping from *names* to *hex specifications*, but ble.sh currently doesn't offer a customization interface to allow that.

**Comment by LeonardoMor** [comment]
Hello, did you succeed in doing this? Looking for something like this but using the tokyonight colors.

**Reply by akinomyoga** [reply]
What specifically would you like to do?  If you would like to configure the color settings, you can use `ble-face`. The list of faces is found on [README §2.6](https://github.com/akinomyoga/ble.sh?tab=readme-ov-file#26-highlight-colors). The detailed descriptions about the graphic specification is found on [Manual §2](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A72-Graphics#2-graphics-settings).

**Reply by LeonardoMor** [reply]
Yes, I was just wondering if someone already made some themes.

**Reply by akinomyoga** [reply]
As far as I know, no color themes (i.e. sets of the `ble-face` settings that are distributed) are made by users. But you might search others' settings on GitHub:

- [Code search results](https://github.com/search?q=-repo%3Aakinomyoga%2Fble.sh+-repo%3Aakinomyoga%2Fblesh-contrib+%2F%5E%5B%5B%3Aspace%3A%5D%5D*ble-face%5Cb%2F+language%3AShell&type=code)

Ah, maybe these are the themes (`bamboo`, `bamboo_light`, and `bamboo_multiplex`):

- [bamboo.nvim/extras/blesh at 5c826c8ad27010ac2fcaf7deb4c36b16d00ef5a1 · ribru17/bamboo.nvim](https://github.com/ribru17/bamboo.nvim/tree/5c826c8ad27010ac2fcaf7deb4c36b16d00ef5a1/extras/blesh)

Note: these seem to use the 24-bit color specification, so you need a terminal that supports 24-bit colors.

**Reply by LeonardoMor** [reply]
Thank you, I was looking for something like this.

**Reply by abhijeeth-babu** [reply]
I tried to make a mocha version using the examples given.

```
# catppuccin mocha
ble-face -s argument_error bg=#f38ba8,fg=#11111b  # Red background, Crust foreground for better contrast
ble-face -s argument_option fg=#f2cdcd,italic  # Flamingo
ble-face -s auto_complete fg=#9399b2,italic  # Overlay2
ble-face -s cmdinfo_cd_cdpath fg=#89b4fa,bg=#11111b,italic  # Blue, Crust
ble-face -s command_alias fg=#b4befe  # Lavender
ble-face -s command_builtin fg=#fab387  # Peach
ble-face -s command_directory fg=#89b4fa  # Blue
ble-face -s command_file fg=#b4befe  # Lavender
ble-face -s command_function fg=#b4befe  # Lavender
ble-face -s command_keyword fg=#cba6f7  # Mauve
ble-face -s disabled fg=#9399b2  # Overlay2
ble-face -s filename_directory fg=#89b4fa  # Blue
ble-face -s filename_directory_sticky fg=#11111b,bg=#a6e3a1  # Crust, Green
ble-face -s filename_executable fg=#a6e3a1,bold  # Green
ble-face -s filename_ls_colors none
ble-face -s filename_orphan fg=#89dceb,bold  # Sky
ble-face -s filename_other none
ble-face -s filename_setgid fg=#11111b,bg=#f9e2af,underline  # Crust, Yellow
ble-face -s filename_setuid fg=#11111b,bg=#fab387,underline  # Crust, Peach
ble-face -s menu_filter_input fg=#11111b,bg=#f9e2af  # Crust, Yellow
ble-face -s overwrite_mode fg=#11111b,bg=#89dceb  # Crust, Sky
ble-face -s prompt_status_line bg=#9399b2  # Overlay2
ble-face -s region bg=#45475a  # Surface1
ble-face -s region_insert bg=#45475a  # Surface1
ble-face -s region_match fg=#11111b,bg=#f9e2af  # Crust, Yellow
ble-face -s region_target fg=#11111b,bg=#cba6f7  # Crust, Mauve
ble-face -s syntax_brace fg=#6c7086  # Overlay0
ble-face -s syntax_command fg=#b4befe  # Lavender
ble-face -s syntax_comment fg=#f9e2af  # Yellow
ble-face -s syntax_delimiter fg=#6c7086  # Overlay0
ble-face -s syntax_document fg=#f5e0dc,bold  # Rosewater
ble-face -s syntax_document_begin fg=#f5e0dc,bold  # Rosewater
ble-face -s syntax_error bg=#f38ba8,fg=#11111b  # Red background, Crust foreground for better contrast
ble-face -s syntax_escape fg=#f2cdcd  # Flamingo
ble-face -s syntax_expr fg=#cba6f7  # Mauve
ble-face -s syntax_function_name fg=#b4befe  # Lavender
ble-face -s syntax_glob fg=#fab387  # Peach
ble-face -s syntax_history_expansion fg=#b4befe,italic  # Lavender
ble-face -s syntax_param_expansion fg=#f38ba8  # Red
ble-face -s syntax_quotation fg=#a6e3a1  # Green
ble-face -s syntax_tilde fg=#cba6f7  # Mauve
ble-face -s syntax_varname fg=#f5e0dc  # Rosewater
ble-face -s varname_array fg=#fab387  # Peach
ble-face -s varname_empty fg=#fab387  # Peach
ble-face -s varname_export fg=#fab387  # Peach
ble-face -s varname_expr fg=#fab387  # Peach
ble-face -s varname_hash fg=#fab387  # Peach
ble-face -s varname_number fg=#f5e0dc  # Rosewater
ble-face -s varname_readonly fg=#fab387  # Peach
ble-face -s varname_transform fg=#fab387  # Peach
ble-face -s varname_unset bg=#f38ba8,fg=#11111b  # Red background, Crust foreground for better contrast
ble-face -s vbell_erase bg=#45475a  # Surface1
```

**Comment by akinomyoga** [comment]
In a commit e4cce0eaa578e0da91d8fd6af7b94a7fbacf13cb, I just added an option [`bleopt color_scheme`](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A72-Graphics#user-content-bleopt-color_scheme) to allow users to select a preset color scheme from the [`contrib/scheme`](https://github.com/akinomyoga/blesh-contrib/tree/master/scheme) directory. Currently, only `default` and `base16` are available, but you may create a PR to add color schemes in the [`blesh-contrib`](https://github.com/akinomyoga/blesh-contrib/pulls) repository.
